### PR TITLE
docs: Update RULES.md to include NOTICES.md information & delete NOTICES.md

### LINKED ISSUE #1006: Improve user experience when interpreting the validation report
## Context
Currently, two files contain information about the rules implemented and the notices emitted by the validator: RULES.md and docs/NOTICES.md. 

#### 1. RULES.md contains the following information:
- The name of the notice 
example: "StartAndEndRangeEqualNotice"
- A small description 
example: "Two date or time fields are equal."
- the severity level (error/warning/info)
- A longer description
example: "The fields frequencies.start_date and frequencies.end_date have been found equal in frequencies.txt. The GTFS spec is currently unclear how this case should be handled (e.g., is it a trip that circulates once?). It is recommended to use a trip not defined via frequencies.txt for this case.)"
- A Reference
example: "Original Python validator implementation" (links to the transitfeed validator)

#### 2. NOTICES.md contains the following information:
- The notice code 
example: start_and_end_range_equal
- The notice schema
example:
Screen Shot 2021-09-09 at 9 31 38 PM

- The files affected
example: frequencies.txt

## Problems
1. Updating those two files create unnecessary manual updates when adding or updating a rule, and often some errors.
2. The user experience when reading the validation result is confusing. 
- step 1: read the notice code in the validation report
- step 2: convert it to the notice name in NOTICES.md
- step 3: read the notice description in RULES.md
- step 4: look at the affected files and the notice schema in NOTICES.md

3. The "Reference" field in RULES.md should be pointing at the specification instead of the transitfeed validator when the rule is validating parts of the spec

## Solution proposed
- Design an ideal user journey mapping and propose changes to the documentation accordingly

## Next steps
- Generate the documentation automatically instead of manually

### LINKED ISSUE #1154: Update references to the python validator in RULES.md
## Problem
Some rules in RULES.md are referencing to Google's original [python validator](https://github.com/google/transitfeed) implementations, whereas others are referencing to the corresponding part of the spec or best practices.
For a casual validator user, the most useful is to link to the spec or best practice mention, in order to have additional context about the issue.
The notices that are only referencing to the python validator should also link to the corresponding spec or best practices, if existing.

## Additional context
https://github.com/MobilityData/gtfs-validator/pull/1132#discussion_r874979905

### PR DESCRIPTION
**Pull request by isabelle-dr**
Closes #1006 
Closes #1154 

**Summary:**
Following the discussions in [PR #1127 (comment)](https://github.com/MobilityData/gtfs-validator/pull/1127#discussion_r862123480), the objective of this PR is to provide a better user experience when reading the validation report. This is part of the effort of the [3.1.0 release](https://github.com/MobilityData/gtfs-validator/issues/1124). 
The HTML report in #1127 will link to the corresponding piece of documentation using the anchor links, in order to ease the process of (1) understanding what the issue is, and (2) fixing the dataset.

This PR contains the following changes:
- the first part of RULES.md has been reworked to make it more useful for a regular user of the validator. The parts that are useful only when contributing have been moved to [NEW_RULES.md](https://github.com/MobilityData/gtfs-validator/blob/documentation-update/docs/NEW_RULES.md), which describes how to add a new rule.
- the information available in NOTICES.md has been added to RULES.md using a collapsable feature to make the file more digestible. The information added is:
   - the notice_codes
   - the notice fields description
   - the affected files
- added the notice code into the first table
- some cleanup & small documentation fixes
- [NEW_RULES.md](https://github.com/MobilityData/gtfs-validator/blob/documentation-update/docs/NEW_RULES.md) has been updated to reflect the change of documentation

**Expected outcomes:** 
- All the documentation that a user needs to understand the validation results & fix their dataset is available in RULES.md
- The user doesn't need to guess the mapping between the CamelCase NoticeName and the snake_case notice_code

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] ~~Run the unit tests with `gradle test` to make sure you didn't break anything~~
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by isabelle-dr** [comment]
@maximearmstrong this is ready for review! :)
Do you know why the tests didn't run on this PR?

File: docs/NEW_RULES.md
Comment (maximearmstrong) [reply docs/NEW_RULES.md:228]: Also, are we not mentioning the infos section because a contributor can't add an info notice? Just wondering.

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:3]:
```suggestion
Note that the notice naming convention changed in `v2` to make contributions of new rules easier by reducing the likelihood of conflicting IDs during parallel development. Please refer to [MIGRATION_V1_V2.md](/docs/MIGRATION_V1_V2.md) for a mapping between v1 and v2 notices.\
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:9]:
```suggestion
A part of the specification that is translated into code in the validator. A Rule will describe if a set of conditions is met or not. For example:
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:10]:
```suggestion
  - In the specification: in the `stops.txt` file, the field `zone_id` is required if providing fare information using `fare_rules.txt` ([source in the spec](https://gtfs.org/schedule/reference/#stopstxt)).
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:15]:
```suggestion
- For example, the output of `StopZoneIdValidator.java` Rule is the Notice `StopWithoutZoneIdNotice`, and it has the notice code `stop_without_zone_id`. 
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:22]:
```suggestion
  * ⚠️ Please note that this validator also generates `System Errors` that give information about things that may have gone wrong during the validation process such as the inability to unzip a GTFS file. These are generated in a second report `system_errors.json`.
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:78]:
```suggestion
| [`StopTimeWithArrivalBeforePreviousDepartureTimeNotice`](#StopTimeWithArrivalBeforePreviousDepartureTimeNotice) | Backwards time travel between stops in `stop_times.txt`                                                                                                | stop_time_with_arrival_before_previous_departure_time  |
```

**Review by maximearmstrong — CHANGES_REQUESTED** [review]
Some suggestions for changes in-line before approval. Thank you for this @isabelle-dr!

File: docs/NEW_RULES.md
Comment (isabelle-dr) [reply docs/NEW_RULES.md:228]: Good catch. The info notice is a bit unclear, we wanted to use it for metadata instead (a bit like you did with the HTML output), because we don't think the current notices are useful to people (still needs to be proven).
But in the meantime, if we want to stay consistent with RULES.md, I think we should add it. Thanks

File: RULES.md
Comment (isabelle-dr) [reply RULES.md:78]: oeil de lynx

**Comment by maximearmstrong** [comment]
> @maximearmstrong this is ready for review! :) Do you know why the tests didn't run on this PR?

Since you are modifying .md files only, the workflows are not running. Most of our workflows are ignored when a PR contains only doc files, this is normal.

**Comment by isabelle-dr** [comment]
Thank you for all these suggestions! It is updated on my end

**Comment by maximearmstrong** [comment]
Thank you @isabelle-dr !

@barbeau I'm ready to approve this PR, but maybe you'd like to do a review first. Let us know :)

**Comment by barbeau** [comment]
@maximearmstrong Thanks, I'll take a look.

File: RULES.md
Comment (barbeau) [inline RULES.md:15]: Related to my below comment on having notice class names in this document - I think including `StopWithoutZoneIdNotice` here is potentially confusing. Why is the name of the notice different from the notice code? If `stop_without_zone_id` is what I see in the validation output (JSON and HTML) why do I need to know the `StopWithoutZoneIdNotice` name? Do I need to reference that somewhere else?

IMHO this file shouldn't contain a bunch of implementation details but should be focused on explaining the business logic of the validation rules. I think the exception is including the collapsed notice field descriptions, which we're doing to try to keep all the docs in the same place.  The fact that we're hiding it by default probably means that most users won't see it (which is fine).

File: RULES.md
Comment (barbeau) [inline RULES.md:99]: Some of these hyperlinks seem to be broken when I click on them - for example:
* [InconsistentAgencyLangNotice](https://github.com/MobilityData/gtfs-validator/blob/6c5a72c58a0cbb8cce4df8f2a96900013400e312/RULES.md#InconsistentAgencyLangNotice)
* [LeadingOrTrailingWhitespacesNotice](https://github.com/MobilityData/gtfs-validator/blob/6c5a72c58a0cbb8cce4df8f2a96900013400e312/RULES.md#LeadingOrTrailingWhitespacesNotice)
* [MissingFeedInfoDateNotice](https://github.com/MobilityData/gtfs-validator/blob/6c5a72c58a0cbb8cce4df8f2a96900013400e312/RULES.md#MissingFeedInfoDateNotice)
* [MissingFeedInfoDateNotice](https://github.com/MobilityData/gtfs-validator/blob/6c5a72c58a0cbb8cce4df8f2a96900013400e312/RULES.md#MissingFeedInfoDateNotice)

...and more below that.

File: RULES.md
Comment (barbeau) [inline RULES.md:19]: Is the double colon here `::` intentional?

File: RULES.md
Comment (barbeau) [inline RULES.md:186]: Maybe `#### Notice JSON fields`? Do these appear anywhere else (outside of the JSON) that a producer/consumer might need to know them?

File: RULES.md
Comment (barbeau) [inline RULES.md:170]: I'm torn as to whether we should include "Affected files", and maybe even "References", in the details section. I think a frequent use case is for the validator to flag an error, a producer or consumer to look up the rule, and then want to verify for themselves that the text in the GTFS spec actually requires/prohibits something - so they want a direct hyperlink to the section of the spec that covers this rule.

Maybe keep "Affected files" outside details, but leave notice JSON fields table and "References" inside details?

**Review by barbeau — COMMENTED** [review]
Thanks @isabelle-dr for working on this! It's definitely nice to get ride of a bunch of the redundant text in NOTICES.md. A few thoughts in-line.

File: RULES.md
Comment (isabelle-dr) [reply RULES.md]: > I believe these anchors for the notice class name that are identical to the Markdown header are redundant

That's true, I removed them in 2a0345600b3e055001d0a822deec7139062a17e9.

For the Class names VS code, I agree with you, this isn't great. Here is how we quickly went about it with @maximearmstrong:
- option 1: replace NoticeClassName by the notice_code in the documentation
  - pros: clear, only one thing to reference a notice
  - cons: 
    - the existing documentation is using NoticeClassName and [some people](https://dot.ca.gov/cal-itp/Critical-GTFS-Validation-Errors#:~:text=has%20deemed%20critical.-,GTFS%20Schedule,-The%20following%20errors) refer to the notices with it.
    - the people looking at RULES.md want to find the description for: (1) the notice, and (2) the notice schema field names. Maybe it's just me, but when I started using the validator, `invalid_phone_number` and `fileName` was the same "thing": a variable that I need to find the description for, so I didn't understand why these had 2 different formats. So from a readability POV it would make sense to have these two things the same format.
- option 2: deprecate notice_code and use NoticeClassName to refer to the notices, and update the JSON report with new code
  - cons: users are currently relying on this code when this validator is integrated in a pipeline, this is a major change.
- option 3 (what we decided to go for right now): keep both for now: add the NoticeClassName in the HTML output and the notice_code in the table at the top of RULES.md
  - pros: users don't have to "guess" which maps to which
  - con: users are confused: why do we have this redundant info? what is notice_code for? Which I tried to mitigate with adding _Notice Code (as it appears in the report.json file)_ in the table 🙈

Does this seem like a reasonable tradeoff? 
This probably deserves more discussion. The most important thing to solve right now is:
- users going back-and-forth between the 2 documents
- users not having the same info in the report VS the documentation

File: RULES.md
Comment (isabelle-dr) [reply RULES.md]: done in ee09f9c923d0f70e6079bdb72c92ed22ba521b27
Kept the first one because it links to the spec and not the validator file

File: RULES.md
Comment (isabelle-dr) [reply RULES.md:14]: > IMHO this file shouldn't contain a bunch of implementation details
Interesting POV. I don't see this as implementation details, but a key element to the comprehension of the validator as a casual user. I will re-work this part to make it less technical details looking!

File: RULES.md
Comment (isabelle-dr) [reply RULES.md]: Okay, after reading your comment & doing a a quick research on how existing users are referencing the notices, I'm also now leaning more towards option 1. I think there is more pro (new users having a better experience) than cons (some existing users might do minor adjustments). @maximearmstrong, if you're OK with this, I will modify this PR to remove class names altogether. 
You would also need to update the HTML output to reflect the change.

Comment (maximearmstrong) [reply RULES.md]: All good for me @isabelle-dr, I'm in favor of option 1 as well.

File: RULES.md
Comment (isabelle-dr) [reply RULES.md:1]: I was thinking we can keep it here since this is the most used file (with the README.md)
Is it a best practice to have all .md files (except README.md) in a /docs folder?

File: RULES.md
Comment (isabelle-dr) [reply RULES.md:186]: In the HTML output it is in a table, I think it's best not to use JSON here for users that only use the HTML output to see the results.
The HTML output:
Screen Shot 2022-05-17 at 9 32 56 AM

The documentation:

Screen Shot 2022-05-17 at 9 58 07 AM

File: RULES.md
Comment (isabelle-dr) [reply RULES.md:170]: Good point. 
I took out the _References_ of the _details_ section to make it more accessible, because for all file requirement notices (such as `empty_column_name`) or best practices (such as `missing_feed_info_date`), the `References` section is more relevant. 
The references to the python validator should be replaced with the reference to the corresponding portion of the spec, but this will have to be in another PR!

**Comment by isabelle-dr** [comment]
I have applied the changes:
- remove notice class to use only notice code
- put the references to the spec outside of the dropdown & add the reference to the spec for notices that were only referring to the python validator
- other small fixes & typos

@barbeau PTAL, the work is done on my end :)

File: RULES.md
Comment (barbeau) [reply RULES.md:1]: I think it could go either way. Normally I'd put all the .md files in the /docs folder (with the exception of README.md). But if you don't want to break existing links to RULES.md I think it's fine to leave it as well.

File: docs/NEW_RULES.md
Comment (barbeau) [inline docs/NEW_RULES.md:6]:
```suggestion
- Notices related to GTFS semantics/business logic are encapsulated within the related validation rule class. See the example below in [`TripUsageValidator`](/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TripUsageValidator.java)
```

File: docs/NEW_RULES.md
Comment (barbeau) [inline docs/NEW_RULES.md:227]:
```suggestion
| [new_rule_related_to_stops](#new_rule_related_to_stops) | new rule short description | new_rule_related_to_stops | 
```

File: RULES.md
Comment (barbeau) [inline RULES.md:10]:
```suggestion
  - In the specification: in the `stops.txt` file, the field `zone_id` is required if providing fare information using `fare_rules.txt` ([source in the specification](https://gtfs.org/schedule/reference/#stopstxt)).
  - ```

File: RULES.md
Comment (barbeau) [inline RULES.md:153]:
```suggestion
### block_trips_with_overlapping_stop_times
```

File: RULES.md
Comment (barbeau) [inline RULES.md:534]:
```suggestion
Value of field with type `email` is not valid. Definitions for valid emails are quite vague. We perform strict validation using the Apache Commons EmailValidator.
```

File: RULES.md
Comment (barbeau) [inline RULES.md:740]:
```suggestion
Value of field with type `url` is not valid. Definitions for valid URLs are quite vague. We perform strict validation using the Apache Commons UrlValidator.
```

**Review by barbeau — COMMENTED** [review]
Thanks for all this work @isabelle-dr! I think it looks much better than before - much easier to read IMHO.

I found a few small typos in the below comments/suggestions.

**Comment by isabelle-dr** [comment]
What an eye @barbeau! 🧐 Thank you

@maximearmstrong, could you have a last look & approve this PR if it all looks good?

File: RULES.md
Comment (barbeau) [inline RULES.md:11]: Looks like the triple ticks were accidentally added in merging a suggestion - this line can be removed.

**Review by barbeau — APPROVED** [review]
LGTM, thanks @isabelle-dr!

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:1432]:
```suggestion
</details>

<a name="EmptyRowNotice"/>
```

File: RULES.md
Comment (maximearmstrong) [inline RULES.md:1455]: For consistency across the document, I'd put the link before the title
```suggestion
<a name="EqualShapeDistanceSameCoordinatesNotice"/>

### equal_shape_distance_same_coordinates
```

**Review by maximearmstrong — APPROVED** [review]
2 small things aside, LGTM! Thank you @isabelle-dr :)

Also, we had to reconfigure the CLA assistant last week and I think we will need your signature in order to merge.

**Comment by maximearmstrong** [comment]
Thank you @isabelle-dr ! The other checks won't be performed because this PR only changes the `.md` files, so I'm merging it.

**Pull request merged**
